recon.py

```diff
@@ -0,0 +1,329 @@
+        """
+        return self._map[mapping].event_type is EventType.TIPTIP
+    
+    def save(self, path: str, metadata: Dict[str, str] = {}):
+        """
+        Save the Reconciliation to path, e.g. ``recon.save('./reconname')``
+        This will save the Reconciliation along with the metadata if metadata
+        is specified, e.g. ``recon.save('./reconname', {'created_at': 'Noon'})``.
+        The metadata is only for book-keeping purposes and is not read when load.
+        """
+        raise NotImplementedError
+    
+    @staticmethod
+    def load(path) -> 'Reconciliation':
+        """
+        Load a Reconciliation from path, e.g. ``Reconciliation.load('./reconname')``.
+        """
+        raise NotImplementedError
+
+class ReconGraph:
+    """
+    ReconGraph is a directed acyclic graph that compactlyrepresent multiple 
+    Reconciliations. The Reconciliation starts with a list of source MappingNode
+    where the parasite first enters the host. Each MappingNode has a list of 
+    corresponding Event. Each event has up to two children MappingNode
+    depending on its type. The sinks of this graph are the TipTip events which
+    have no children MappingNode.
+    """
+    def __init__(self, sources: List[MappingNode], initial_map: Dict[MappingNode, List[Event]] = {}):
+        """
+        Creates a ReconGraph. The ReconGraph starts with source
+        MappingNode where the parasite first enters the host.
+        """
+        self.sources = sources
+        self._map = initial_map
+    
+    def __eq__(self, other):
+        return type(self) == type(other) and self._map == other._map
+    
+    def __repr__(self):
+        return "%s(sources=%s, %s)" % (type(self).__name__, self.sources, self._map)
+    
+    def add_event(self, mapping: MappingNode, event: Event):
+        """
+        Add an event corresponding to mapping.
+        """
+        if mapping not in self._map:
+            self._map[mapping] = []
+        self._map[mapping].append(event)
+
+    def set_events(self, mapping: MappingNode, events: List[Event]):
+        """
+        Set a mapping to correspond to a list of events.
+        """
+        self._map[mapping] = events
+    
+    def events_of(self, mapping: MappingNode) -> List[Event]:
+        """
+        Get a list of events corresponding to mapping.
+        """
+        return self._map[mapping].copy()
+    
+    def is_sink(self, mapping: MappingNode) -> bool:
+        """
+        Whether the host and the parasite of the mapping are leaves of its
+        phylogenic tree with a tip mapping.
+        """
+        return len(self._map[mapping]) == 1 and self._map[mapping][0].event_type is EventType.TIPTIP
+    
+    def enumerate(self) -> List[Reconciliation]:
+        """
+        Returns all Reconciliation represented in this ReconGraph.
+        """
+        raise NotImplementedError
+
+    def save(self, path: str, metadata: Dict[str, str] = {}):
```

Comment:
I'm planning to save it as json because it's easy to look at the structure in an editor.